### PR TITLE
Make random values collision detection aware of conditional params

### PR DESCRIPTION
**Pull request by omalleyt12**
* As part of this PR, factors random values generation into the base oracle
* Also increased the number of restarts in the `BayesianOptimization` oracle because the change to use a Matern kernel with nu=2.5 allows `scipy.minimize` to converge faster but seems to require more restarts to work well

### CODE REVIEW

File: kerastuner/engine/oracle.py [hunk 1]
```diff
@@ -89,6 +91,13 @@ def __init__(self,
         self.directory = None
         self.project_name = None
 
+        # Random trial generation attributes.
+        self.seed = seed or random.randint(1, 1e4)
```
Comment (fchollet) [inline kerastuner/engine/oracle.py]: For layers, for instance, if the seed is passed as `seed=None`, it would get set as `self.seed = None` (and also saved as None in the config) and thus the layer would start with a new seed if recreated from the same config.

Is this also something we should do here, or do we believe that saving the seed that's generated on the fly is  better?

File: kerastuner/engine/oracle.py [hunk 2]
```diff
@@ -333,11 +349,37 @@ def _get_oracle_fname(self):
...
+    def _compute_values_hash(self, hps):
+        # Don't include conditional params that are not currently active.
+        values = {name: val for name, val in hps.values.items()
+                  if hps.is_active(name)}
         keys = sorted(values.keys())
         s = ''.join(str(k) + '=' + str(values[k]) for k in keys)
         return hashlib.sha256(s.encode('utf-8')).hexdigest()[:32]
 
+    def _random_populate_space(self, allow_collisions=False):
+        for _ in range(self._max_collisions):
```
Comment (fchollet) [inline kerastuner/engine/oracle.py]: Perhaps this would read a bit more intuitively with

```
while collisions < self._max_collisions:
    ...
    if self._compute_values_hash(trial_hps) in self._tried_so_far:
        collisions += 1  # no need for comment since the meaning is transparent
```

File: kerastuner/engine/oracle.py [hunk 3]
```diff
@@ -333,11 +349,37 @@ def _get_oracle_fname(self):
...
+                # Attempt to avoid collisions.
+                continue
+            return {'status': trial_lib.TrialStatus.RUNNING,
+                    'values': trial_hps.values}
+
+        # Max collisions occurred.
+        if allow_collisions:
+            return {'status': trial_lib.TrialStatus.RUNNING,
+                    'values': trial_hps.values}
+        # Search space is presumed exhausted.
```
Comment (fchollet) [inline kerastuner/engine/oracle.py]: The number of collisions is fixed at 20, but 20 is too much if there are 5 possible values and too little if there are 2000.

Perhaps this should be instead a function of the size of the search space (i.e. not an instance attribute but something set at the start of `_random_populate_space`, e.g. 20% of the number of possible values). Exact collisions are only possible for Int, Bool, or Choice, and for each of these the number of values is provided by the constructor params.

Another possibility is to randomly iterate over the values of discrete hps, so that the full space is guaranteed to be covered when it is discrete.

Perhaps this is overly complex, though.

File: kerastuner/tuners/bayesian.py [hunk 4]
```diff
@@ -114,7 +105,7 @@ def _upper_confidence_bound(x):
 
         optimal_val = float('inf')
         optimal_x = None
-        num_restarts = 50
+        num_restarts = 150
```
Comment (fchollet) [inline kerastuner/tuners/bayesian.py]: This would be worth a comment.

File: kerastuner/engine/hyperparameters.py [hunk 5]
```diff
@@ -483,6 +483,21 @@ def conditional_scope(self, parent_name, parent_values):
...
+
+        # Arguments:
+          name: The name of the hyperparameter.
+
+        # Returns:
+          Whether the hyperparameter is active for the current trial.
+        """
+        return self._get(name, error_on_inactive=False) is not None
+
     def _conditions_are_active(self, scopes=None):
```
Comment (yogi-dmt) [inline kerastuner/engine/hyperparameters.py]: Is there a way to make this work with loops ie. multiple layers? It seems like to apply this to layers we need to write out each layer and have an conditional hyper param for each? Is there a way for the program to automatically detect if a param is active or not without having to supply the conditional HP?

**Comment by yogi-dmt** [comment]
Since this only works for when you use the conditional_scope hyperparameter, i'm not sure this does very well with loops. Instead of trying to manually define when a param is active or not, maybe it is better to just build the model and keep a list of active params on the hyperparameters attrib when we retrieve a value? Then compute the hash. Also would help if there is a future way to make bayesian optimization conditional.

**Comment by omalleyt12** [comment]
This PR is superceded by another PR that reimplements conditional hps
